### PR TITLE
dec: refactor raw parts + 0.4.2 release

### PR DESCRIPTION
**Pull request by sploiselle**

### CODE REVIEW

File: dec/src/decimal.rs [hunk 1]
```diff
@@ -200,6 +195,16 @@ impl<const N: usize> Decimal<N> {
...
 
+    /// Returns the number of elements required in the `lsu` to represent some
+    /// number of digits.
+    ///
+    /// This function is public and accepts a `u32` instead of a `Decimal` to
+    /// aid in decomposing ([`Self::to_raw_parts`]) and recomposing
+    /// ([`Self::from_raw_parts`]) values.
+    pub fn digits_to_lsu_elements_len(digits: u32) -> usize {
+        (usize::try_from(digits).unwrap() + decnumber_sys::DECDPUN - 1) / decnumber_sys::DECDPUN
+    }
```
Comment (benesch) [inline dec/src/decimal.rs:206]: This feels like something that doesn't need to be public? Isn't this only needed by the asserts in `to_raw_parts` and `from_raw_parts`?

File: dec/src/decimal.rs [hunk 2]
```diff
@@ -313,24 +311,58 @@ impl<const N: usize> Decimal<N> {
...
+    /// If `lsu_u8` is not a slice that can be recomposed into a `&[16]` with
+    /// the number of digits implicitly specified by the `digits` parameter,
+    /// i.e. essentially `ceil(digits / 3)`. You can determine the appropriate
+    /// number of elements in `lsu_u8` using 2 *
+    /// [`Decimal::digits_to_lsu_elements_len`].
+    pub unsafe fn from_raw_parts(digits: u32, exponent: i32, bits: u8, lsu_u8: &[u8]) -> Self {
+        let lsu_u16: &[u16] = std::slice::from_raw_parts(
+            lsu_u8.as_ptr() as *const u16,
+            lsu_u8.len() / (std::mem::size_of::<u16>() / std::mem::size_of::<u8>()),
+        );
```
Comment (benesch) [inline dec/src/decimal.rs:357]: So this is the part that is sketching me out! I don't think it's valid to construct a `[u16]` out of a `[u8]` because of alignment requirements. At least I'm pretty sure it's UB. But going in the opposite direction is definitely safe:

```rust
    const N: usize = 4;
    let mut lsu = [0_u16; N];
    let lsu_bytes = unsafe {
        slice::from_raw_parts_mut(
            lsu.as_mut_ptr() as *mut u8,
            lsu.len() * mem::size_of::<u16>(),
        )
    };
    lsu_bytes.copy_from_slice(&[1, 2, 3, 4, 5, 6, 7, 8]);
```

tl;dr if you convert the `[u16]` into a `[u8]`, and then copy into _that_, you're definitely good, because it's always valid to go from `[u16]` to `[u8]`.

File: dec/src/decimal.rs (hunk 1 above)
Comment (sploiselle) [reply dec/src/decimal.rs:206]: If you take the output of `to_raw_parts` and blat them down somewhere else, e.g. in a row, this function gives you a way to determine the number of bytes you need to read out of the row to provide valid input to `from_raw_parts` as a function of `digits`, i.e. this is used on my fork of MZ locally for deserializing rows back to `Datum::APD`.

This doesn't need to be exported from the library, but we would need to have the same logic somewhere else; figured this was the most ergonomic place and addressed the oddity in the doc comment.

File: dec/src/decimal.rs (hunk 2 above)
Comment (sploiselle) [reply dec/src/decimal.rs]: Ah––I really learned a lot here. I clearly didn't understand the alignment issue, but I've since closely read the docs and come up with some test cases that fail in the old implementation. ty for your patience and hope this moves the PR forward.

File: dec/src/decimal.rs (hunk 1 above)
Comment (benesch) [reply dec/src/decimal.rs:206]: Oh, d'oh, right, sounds good.

File: dec/src/decimal.rs [hunk 3]
```diff
@@ -313,24 +311,61 @@ impl<const N: usize> Decimal<N> {
...
         let mut lsu = [0; N];
-        lsu[..lsu_in.len()].copy_from_slice(lsu_in);
+        for (i, v) in lsu_u8.chunks(2).enumerate() {
+            // Use native endian ordering, which matches the ordering generated
+            // for the `lsu`/`coefficient_units()` in `to_raw_parts`.
+            lsu[i] = u16::from_ne_bytes(
+                (*v).try_into()
+                    .expect("converting slice of 2 u8s to [u8; 2] to succeed"),
+            );
+        }
```
Comment (benesch) [inline dec/src/decimal.rs:368]: Nice safe solution! And if this ever shows up in profiles we can evaluate then if a little `unsafe` would be worthwhile.

**Review by benesch — APPROVED** [review]
This looks great!
